test/e2e/storage/testsuites/provisioning.go

```diff
@@ -283,11 +313,54 @@ func TestDynamicProvisioning(t StorageClassTest, client clientset.Interface, cla
 			command += fmt.Sprintf(" && ( mount | grep 'on /mnt/test' | awk '{print $6}' | sed 's/^(/,/; s/)$/,/' | grep -q ,%s, )", option)
 		}
 		command += " || (mount | grep 'on /mnt/test'; false)"
-		runInPodWithVolume(client, claim.Namespace, claim.Name, t.NodeName, command, t.NodeSelector, t.ExpectUnschedulable)
+
+		// This test uses TestConfig.ClientNodeName and TestConfig.ClientNodeSelector
+		// as follows:
+		// - first it runs two pods on the same node, selected based on
+		//   TestConfig.ClientNodeName (if not empty) or TestConfig.ClientNodeSelector
+		//   (otherwise)
+		// - if TestConfig.ClientNodeSelector matches more than the node from
+		//   the first step, it will run a third pod on a different node
+
+		By("checking the created volume is writable")
+		runInPodWithVolume(client, claim.Namespace, claim.Name, "-first", nodeName, command,
+			nil /* node selector */, false /* expect unschedulable */)
 
 		By("checking the created volume is readable and retains data")
-		runInPodWithVolume(client, claim.Namespace, claim.Name, t.NodeName, "grep 'hello world' /mnt/test/data", t.NodeSelector, t.ExpectUnschedulable)
+		runInPodWithVolume(client, claim.Namespace, claim.Name, "-second", nodeName, "grep 'hello world' /mnt/test/data", nil /* node selector */, t.ExpectUnschedulable)
+
+		// Run on another node, if we have one.
```

Comment:
Picking the 2nd node should still respect any node selectors set.  For example, for GCE PD, the 2nd node needs to be in the same zone as the 1st node.

Reply:
The two test pods run on the exact same node, and that node is guaranteed to come from the nodes selected for the test via the node selector. Isn't that what you are asking for?

Reply:
Perhaps I need to point out that TestDynamicProvisioning never (?!) ran with a non-zero `t.NodeSelector`. `t.NodeSelector` itself was introduced fairly recently (https://github.com/kubernetes/kubernetes/commit/12ece9dacbbe8b3f9727c5136912de9089b54b27#diff-6dc833c65b9c544adfb2d654d2f7cbd3) and even after that commit, `testProvisioning` did not set it (https://github.com/kubernetes/kubernetes/blob/12ece9dacbbe8b3f9727c5136912de9089b54b27/test/e2e/storage/testsuites/provisioning.go#L87-L100).

As far as I can tell, `t.NodeSelector` was introduced for the `TestBinding` tests and is only used there.

In other words, my changes shouldn't cause the test to stop using a node selector.